### PR TITLE
ONNX Update training ops and training amenable export API

### PR DESCRIPTION
**Pull request by lara-hdr**
- Update Dropout and Batchnorm in opset 12 : https://github.com/onnx/onnx/pull/2568
- Update api logic for exporting to ONNX training amenable models

### CODE REVIEW

File: torch/onnx/symbolic_opset12.py [hunk 1]
```diff
@@ -9,7 +10,56 @@
 
 # This file exports ONNX ops for opset 12
 
+black_listed_operators = [
```
Comment (lara-hdr) [inline torch/onnx/symbolic_opset12.py]: unrelated to this PR; making sure to blacklist updated ops in opset 12 that are not updated in PyTorch yet.
Will submit a separate PR to implement those.

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 55f46922:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=55f46922d706217da621cf2ede94c7f86aa7e93c).

### 1 failure *not* recognized by patterns:
| Job | Step | Status |
| --- | --- | --- |
| CircleCI <sup>binary_linux_libtorch_2_7m_cpu_devtoolset7_shared-with-deps_build</sup> | <sup>Checkout pytorch/builder repo</sup> | New in PR |


---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: torch/onnx/utils.py [hunk 2]
```diff
@@ -45,7 +63,7 @@ def set_training(model, mode):
             model.train(old_mode)
 
 
-def export(model, args, f, export_params=True, verbose=False, training=False,
```
Comment (lara-hdr) [inline torch/onnx/utils.py]: "training" param already exists, but we do not support exporting fully trainable params. 
Would it be ok to rename this param to something like "training_amenable" or "training_friendly" to avoid user confusion?

File: torch/onnx/symbolic_opset12.py (hunk 1 above)
Comment (spandantiwari) [reply torch/onnx/symbolic_opset12.py]: Yes, let's do that.

File: torch/onnx/utils.py (hunk 2 above)
Comment (spandantiwari) [reply torch/onnx/utils.py]: @houseroad - is the existing `training` parameter used anywhere? Can we consider removing it, so as to avoid confusion?

File: torch/onnx/symbolic_helper.py [hunk 3]
```diff
@@ -461,6 +461,11 @@ def _set_operator_export_type(operator_export_type):
     global _operator_export_type
     _operator_export_type = operator_export_type
 
+_training_mode = None
+def _set_training_mode(training_mode):
+    global _training_mode
```
Comment (spandantiwari) [inline torch/onnx/symbolic_helper.py]: Do we have to make this global. Can we scope it more to a file and then import it where ever necessary?

File: torch/onnx/utils.py [hunk 4]
```diff
@@ -32,20 +32,42 @@ def is_in_onnx_export():
...
+
+        if mode is None:
+            mode = False
+            # if the model is in training mode but the user did not specify
+            # to export the model in training mode, export the model in inference
+            # mode (default) and warn them
+            if old_mode is True:
+                warnings.warn("You are exporting the model to ONNX while in training mode with "
+                              "train=None param. The model will default to inference mode export. "
+                              "If you wish to export a training amenable ONNX model, specify train=True "
```
Comment (spandantiwari) [inline torch/onnx/utils.py]: nit: we should update training amenable to "training-amenable" everywhere.

File: torch/onnx/utils.py [hunk 5]
```diff
@@ -32,20 +32,42 @@ def is_in_onnx_export():
...
+                              "in torch.onnx.export().")
+
+        from torch.onnx.symbolic_helper import _set_training_mode
+        _set_training_mode(mode)
+        # ONNX opset 12 has better support for training amenable models, with updated
+        # versions of the dropout and batch_norm operators
+        if mode is True:
+            from torch.onnx.symbolic_helper import _export_onnx_opset_version
+            if _export_onnx_opset_version < 12:
+                warnings.warn("You are exporting the model in training mode with onnx opset version {}. "
```
Comment (spandantiwari) [inline torch/onnx/utils.py]: I would be word the warning stronger to something like, "Opsets lower than Opset 12 will not be able to able to export nodes such as Dropout and bn correctly.".

File: torch/onnx/utils.py [hunk 6]
```diff
@@ -467,80 +489,81 @@ def _export(model, args, f, export_params=True, verbose=False, training=False,
...
-                else zipfile.ZIP_STORED
-            with zipfile.ZipFile(f, 'w', compression=compression) as z:
-                z.writestr(ONNX_ARCHIVE_MODEL_PROTO_NAME, proto)
-                for k, v in export_map.items():
-                    z.writestr(k, v)
-        elif export_type == ExportTypes.DIRECTORY:
-            import os
-            if os.path.exists(f):
-                assert(os.path.isdir(f))
+        with set_training(model, training):
```
Comment (spandantiwari) [inline torch/onnx/utils.py]: Since this encapsulating most of the method body, we should consider a different name, such as `select_model_mode` or `select_training_vs_eval_mode`. The word `set` may give an impression that the mode is set for ever, and not the fact that it is only temporarily changed.

Comment (spandantiwari) [reply torch/onnx/utils.py]: Also, a brioef comment here (also) about what `set_training` is doing would be nice.

File: torch/onnx/utils.py [hunk 7]
```diff
@@ -467,80 +489,81 @@ def _export(model, args, f, export_params=True, verbose=False, training=False,
...
+                    val_add_node_names, val_use_external_data_format, model_file_location)
             else:
-                os.makedirs(f)
+                proto, export_map = graph._export_onnx(
+                    {}, opset_version, dynamic_axes, False, operator_export_type,
+                    strip_doc_string, val_keep_init_as_ip, custom_opsets, val_add_node_names,
+                    val_use_external_data_format, model_file_location)
+
+            if enable_onnx_checker and \
+                operator_export_type is OperatorExportTypes.ONNX_ATEN_FALLBACK and \
```
Comment (spandantiwari) [inline torch/onnx/utils.py]: FYI - these lines may need a manual merge with latest master.

File: torch/onnx/symbolic_opset12.py [hunk 8]
```diff
@@ -9,7 +10,56 @@
...
+
+@parse_args('v', 'v', 'v', 'v', 'v', 'i', 'f', 'f', 'i')
+def batch_norm(g, input, weight, bias, running_mean, running_var, training, momentum, eps, cudnn_enabled):
+    input_sizes = input.type().sizes()
+
+    if weight is None or sym_help._is_none(weight):
+        assert len(input_sizes) > 1
+        weight_value = torch.tensor([1.] * input_sizes[1]).type(
+            'torch.' + input.type().scalarType() + 'Tensor')
+        weight = g.op("Constant", value_t=weight_value)
```
Comment (spandantiwari) [inline torch/onnx/symbolic_opset12.py]: Is this a parameter that is expected to be trainable in the training-amenable mode? If yes, then if these are going to be translated to onnx::constant, then that would be a problem. We should try and keep all trainable parameters as initializers, because onnx::constant would mean that they are immutable.

File: torch/onnx/symbolic_opset9.py [hunk 9]
```diff
@@ -1252,7 +1252,8 @@ def exp(g, self):
 
 @parse_args('v', 'f', 'i')
 def dropout(g, input, p, train):
-    if not train:  # in eval mode, dropout is non-op
+    # in eval mode, dropout is non-op - if the node's train param is set to False, dropout is non-op
+    if not sym_help._training_mode:
         return input
     warnings.warn("Dropout is a training op and should not be exported in inference mode. "
```
Comment (spandantiwari) [inline torch/onnx/symbolic_opset9.py]: Can we update this warning to be more informative about how we cannot export Dropout in training mode for opsets < 12?

File: test/onnx/test_operators.py [hunk 10]
```diff
@@ -16,6 +16,8 @@
 import shutil
 import torch.testing._internal.common_utils as common
 
+import unittest
```
Comment (houseroad) [inline test/onnx/test_operators.py]: nit: from unittest import skip?

File: torch/onnx/utils.py (hunk 2 above)
Comment (houseroad) [reply torch/onnx/utils.py]: I am not aware of any use case.
`training` is aligned with parameter in pytorch model.

We can turn it into an enum parameter: 1) Eval, 2) Preserve, 3) Train. I prefer to use eval as default.

**Review by houseroad — COMMENTED** [review]
I think it makes sense to make eval as default mode for onnx export. Since most of use cases, exported model should be inference, and we don't want to change the default behavior to keep the backward compatibility.

File: scripts/onnx/test.sh [hunk 11]
```diff
@@ -62,5 +63,6 @@ if [[ "$BUILD_ENVIRONMENT" == *py3* ]]; then
   pytest "${args[@]}" \
     "$top_dir/test/onnx/test_pytorch_onnx_onnxruntime.py" \
     "$top_dir/test/onnx/test_custom_ops.py" \
-    "$top_dir/test/onnx/test_models_onnxruntime.py"
+    "$top_dir/test/onnx/test_models_onnxruntime.py" \
+    "$top_dirtest/test/onnx/test_utility_funs.py"
```
Comment (BowenBao) [inline scripts/onnx/test.sh]: nit: typo?

Comment (lara-hdr) [reply scripts/onnx/test.sh]: thanks!!

File: test/onnx/test_utility_funs.py [hunk 12]
```diff
@@ -325,6 +328,95 @@ def test_error_on_data_parallel(self):
...
+                          opset_version=self.opset_version, training=torch.onnx.TrainingMode.TRAINING)
+        ort_sess = onnxruntime.InferenceSession(f.getvalue())
+
+        ort_inputs = {ort_sess.get_inputs()[0].name : x.cpu().numpy()}
+        ort_outs = ort_sess.run(None, ort_inputs)
+        [np.testing.assert_allclose(p_out, ort_out, atol=10e-3, rtol=10e-3) for p_out, ort_out in zip(pytorch_out, ort_outs)]
+
+    # TODO: Enable test when Dropout is implemented in ORT for opset 12.
+    @skipIfUnsupportedOpsetVersion([12])
+    def test_dropout_training(self):
```
Comment (BowenBao) [inline test/onnx/test_utility_funs.py]: Are we testing these for opset < 12?

File: torch/onnx/symbolic_opset9.py [hunk 13]
```diff
@@ -1019,8 +1019,8 @@ def batch_norm(g, input, weight, bias, running_mean, running_var, training, mome
     out = g.op("BatchNormalization", input, weight, bias, running_mean, running_var,
                epsilon_f=eps,
                momentum_f=1 - momentum,
-               outputs=1 if not training else 5)
-    if not training:
+               outputs=1 if not sym_help._training_mode else 5)
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py]: For here and below, why are we switching from using `training`?

Comment (lara-hdr) [reply torch/onnx/symbolic_opset9.py]: since we have a training parameter now it would be more consistent to use it everywhere.
Do you think it would be better to keep using the input 'training' instead?
Maybe I could add an assertion that they are equal here?

File: test/onnx/test_utility_funs.py (hunk 12 above)
Comment (lara-hdr) [reply test/onnx/test_utility_funs.py]: tests are in test_operators.py and test_utility_funs.py

File: torch/onnx/symbolic_opset9.py (hunk 13 above)
Comment (BowenBao) [reply torch/onnx/symbolic_opset9.py]: I see, I agree we should use the new parameter for consistency. Let's add a helper function asserting that `training` input is equivalent to training parameter.

Comment (lara-hdr) [reply torch/onnx/symbolic_opset9.py]: @BowenBao @spandantiwari,

For context: previously we were using the param "training" from the inputs of the symbolics of dropout and bn. After adding the attribute _training_mode, we are now using this parameter instead for consistency.

To address this comment, I added a function in symbolic_helper to assert that the "training" param for bn and dropout has the same value as the new _training_mode attribute.

I saw failures happening in test_operators.py; and after debugging, I realized that with lambdas the "training" param does not behave as expected... In that case the model is a FuncModule() with a ParameterList(), and setting the model's mode does not modify the value of "training" in bn and dropout.

In that case, we would want to trust the _training_mode param rather than the the training param given in the symbolic.
I changed the assert to throw a warning instead, we would then use the value of _training_mode for the model's mode, but wanted to consult you in case you have a better idea of how to deal with this situation.

File: test/onnx/test_models.py [hunk 14]
```diff
@@ -37,9 +37,10 @@ def toC(x):
 
 class TestModels(TestCase):
     def exportTest(self, model, inputs, rtol=1e-2, atol=1e-7):
-        graph = torch.onnx.utils._trace(model, inputs, OperatorExportTypes.ONNX)
-        torch._C._jit_pass_lint(graph)
-        verify(model, inputs, backend, rtol=rtol, atol=atol)
+        with torch.onnx.select_model_mode_for_export(model, False):
```
Comment (BowenBao) [inline test/onnx/test_models.py]: What is `False`? Should it be `None`?

File: torch/onnx/utils.py [hunk 15]
```diff
@@ -281,15 +304,14 @@ def _trace_and_get_graph_from_model(model, args, training):
     # before and after running the model.  Fail fast!
     orig_state_dict_keys = _unique_state_dict(model).keys()
 
-    # By default, training=False, which is good because running a model in
+    # By default, training=TrainingMode.EVAL, which is good because running a model in
```
Comment (BowenBao) [inline torch/onnx/utils.py]: The comment is contradicting the parameter above `training=None`

File: torch/onnx/symbolic_helper.py [hunk 16]
```diff
@@ -413,6 +413,16 @@ def _avgpool_helper(tuple_fn, padding, kernel_size, stride, divisor_override, na
...
 
+def assert_training_mode(op_mode, op_name):
+    global _training_mode
+    op_mode = True if op_mode == 1 else False
+    if op_mode != _training_mode:
+        op_mode = "training " if op_mode else "inference"
+        training_mode = "training " if _training_mode else "inference"
+        warnings.warn("ONNX export mode is set to " + training_mode +
+                      " mode, but operator " + op_name + " is exporting in " +
+                      op_mode + " mode.")
```
Comment (BowenBao) [inline torch/onnx/symbolic_helper.py]: Should we add another line stating that `op_mode` is ignored?

File: torch/onnx/utils.py [hunk 17]
```diff
@@ -281,15 +304,14 @@ def _trace_and_get_graph_from_model(model, args, training):
...
     # training mode could result in internal buffers getting updated, dropout
     # getting applied, etc.  If you really know what you're doing, you
-    # can turn training=True (or None, to preserve whatever the original
-    # training mode was.)
-    with set_training(model, training):
```
Comment (BowenBao) [inline torch/onnx/utils.py]: We are removing `set_training` here. This makes `training` in the method an unused parameter. Should it be remove from this method?

File: torch/onnx/utils.py [hunk 18]
```diff
@@ -298,7 +320,7 @@ def _trace_and_get_graph_from_model(model, args, training):
     return trace_graph, torch_out
 
 
-def _model_to_graph(model, args, verbose=False, training=False,
+def _model_to_graph(model, args, verbose=False, training=None,
```
Comment (BowenBao) [inline torch/onnx/utils.py]: the same applies here as well. `training` parameter is not really used inside the method. It all depends on the caller of `_model_to_graph` to properly set training mode, by calling `with select_model_mode_for_export(model, training):`, otherwise results may be unexpected.

**Review by BowenBao — APPROVED** [review]

**Review by houseroad — COMMENTED** [review]
Could you rebase to master and trigger the test again?

**Comment by houseroad** [comment]
https://github.com/pytorch/pytorch/pull/32950/checks?check_run_id=512600263 clang-tidy is broken, could you fix it?

**Comment by lara-hdr** [comment]
@houseroad it seems like after adding the "Training" enum, clag-tidy is trying to include onnx-ml.pb.h (instead of onnx.ph.h) which is the reason of the failure.  

I am not sure why this is happening since ONNX_ML is not defined..
https://github.com/onnx/onnx/blob/9fdae4c68960a2d44cd1cc871c74a6a9d469fa1f/onnx/onnx_pb.h#L50

File: .github/workflows/lint.yml [hunk 19]
```diff
@@ -198,6 +198,7 @@ jobs:
             -g"-torch/csrc/jit/export.cpp"            \
             -g"-torch/csrc/jit/import.cpp"            \
             -g"-torch/csrc/jit/netdef_converter.cpp"  \
+            -g"-torch/csrc/onnx/init.cpp"             \
```
Comment (lara-hdr) [inline .github/workflows/lint.yml]: @houseroad added this line to exclude init.cpp (which includes onnx_pb.h) to fix the linter error, as explained in the comment above (line 191-193)

**Comment by lara-hdr** [comment]
cc @houseroad clang-tidy issue is fixed
